### PR TITLE
[4.x] Add a default cURL HTTP client

### PR DESCRIPTION
**Pull request by cleptric**
**_Work in Progress_**

This adds a cURL-based HTTP client and simplifies the setup when using a custom `Transport` or `HttpClient`.
As mentioned in #1586, this removes the `/store` endpoint entirely. All events are now sent to the `/envelope` endpoint.

### CODE REVIEW

File: src/ClientBuilder.php
Comment (Jean85) [inline src/ClientBuilder.php:173]: Having `getClient` and `getHttpClient` may be confusing... What's the difference? Can we use another name?

We also have Transport too!

File: src/HttpClient/HttpClient.php
Comment (Jean85) [inline src/HttpClient/HttpClient.php:39]: What happens when we set the DSN to `null` to disable the client? This may be a big BC, for framework libraries too, we need to propagate this outside.

File: src/HttpClient/HttpClient.php
Comment (Jean85) [inline src/HttpClient/HttpClient.php:47]: What if I want to disable this? i.e. if I have a local, self-hosted, not exposed, not behind TLS Sentry server?

[edit] I saw the TODO below, this row is a duplicate.

File: src/HttpClient/HttpClient.php
Comment (Jean85) [inline src/HttpClient/HttpClient.php:48]: Isn't this the option to have compression? We may skip that in this PR (gzip caused issues in the past) but we may require it as a feature later.

File: src/HttpClient/HttpClient.php
Comment (Jean85) [inline src/HttpClient/HttpClient.php:116]: These two values always go together... Maybe they need a dedicated value object?

File: src/HttpClient/HttpClient.php
Comment (Jean85) [inline src/HttpClient/HttpClient.php:140]: This is a BIG hot spot: making some mistakes here would open the door for CVEs.

File: src/HttpClient/Response.php
Comment (Jean85) [inline src/HttpClient/Response.php:62]: This is fragile: what if we can't fetch the error message?

**Review by Jean85 — COMMENTED** [review]
As I've already said in https://github.com/getsentry/sentry-php/issues/1586#issuecomment-1721864613, I'm against writing our own HTTP client but, since I recognize I don't have a final say in this, I'm still giving my review.

I would at least suggest leveraging PSR-18 (HTTP client) and so PSR-7 (request/response) with it, but we could ditch PSR-17 (factories) to keep it simple.

**Comment by ste93cry** [comment]
I agree with everything already said by @Jean85: first of all, I don’t understand the reason to have our own `HttpClientInterface` instead of just using the PSR-18 one. There is no benefit I can think of, it would be just another layer that needs to be implemented by the end user. Second, and more importantly, I don’t understand why we are reimplementing a full cURL client, which we have to maintain and for which there is no benefit because it’s a lot simpler and less customizable than any other out of here. Moreover, what’s happening here is that we are reinventing the wheel: I can understand that this SDK has to ship with a default client, even though I don’t agree, but at least choose one reliable client instead of reinventing the wheel. We had our own HTTP client in the past, and it was full of issues. Given that the SDK is integrated with both Laravel and Symfony, it would be wise to choose the `symfony/http-client` which is common for both

File: src/Transport/Result.php
Comment (ste93cry) [inline src/Transport/Result.php:11]: What's the point of removing `final` here? This class is a value object and it doesn't need to be extended by third parties

File: src/Transport/ResultStatus.php
Comment (ste93cry) [inline src/Transport/ResultStatus.php:11]: What's the point of removing `final` here? This class represents an `enum`, which by definition has a fixed set of values, so it doesn't have to be extended or it would be pointless

File: src/Transport/ResultStatus.php
Comment (ste93cry) [inline src/Transport/ResultStatus.php:11]: What's the point of removing `final` here? This class represents an enum, which by definition has a fixed set of values, so it doesn't have to be extended or it would be pointless

File: src/HttpClient/Response.php
Comment (ste93cry) [inline src/HttpClient/Response.php:7]: What's the point of removing `final` here? This class is a value object and it doesn't need to be extended by third parties

File: src/HttpClient/HttpClient.php
Comment (cleptric) [reply src/HttpClient/HttpClient.php:48]: This flag only handles automatic decompression of HTTP downloads. Setting it to `''` enables all supported built-in compressions.

File: src/HttpClient/HttpClient.php
Comment (cleptric) [reply src/HttpClient/HttpClient.php:140]: Please expand on this. The only two header values currently used are `X-Sentry-Rate-Limits` and `Retry-After` and I fail to see how this could lead to a vulnerability.

File: src/HttpClient/HttpClient.php
Comment (cleptric) [reply src/HttpClient/HttpClient.php:39]: An upcoming feature will require us always to construct an envelope in certain scenarios. I agree that this is not the best solution. I have to think about improving it.

File: src/HttpClient/Response.php
Comment (cleptric) [reply src/HttpClient/Response.php:62]: https://github.com/getsentry/sentry-php/blob/fa5145d9d7f0531291769b2296272ca4c57cf412/src/HttpClient/HttpClient.php#L86 currently sets this. All other cases pass `''`. We could make it optional, but I don't see a scenario were this would not work.

File: src/HttpClient/HttpClient.php
Comment (cleptric) [reply src/HttpClient/HttpClient.php]: I added `http_ssl_verify_peer`. Your use case is the same one I had in mind.

File: src/ClientBuilder.php
Comment (Jean85) [reply src/ClientBuilder.php:173]: Should we mark that as `@internal` then?

Comment (cleptric) [reply src/ClientBuilder.php:173]: Totally, this is all still WIP, as mentioned in the PR description. 😃
Docs and annotation will follow once the implementation and tests are in place.

File: src/HttpClient/HttpClient.php
Comment (Jean85) [reply src/HttpClient/HttpClient.php:140]: Just today I got a notification about a Guzzle CVE that hit them a few months ago with a 7.5 score about headers: https://security.snyk.io/vuln/SNYK-PHP-GUZZLEHTTPPSR7-5426094

Handling headers could open us to CVE reports.

Comment (cleptric) [reply src/HttpClient/HttpClient.php:140]: Yep, but I would argue that this heavily depends on the context of how the header values are used. We could limit the parsing on the mentioned two headers. In the context of Guzzle, I agree that this is much more critical than in our case.

Most SDKs will talk to a somewhat "trusted" server, be it Sentry SaaS or self-hosted, and even if someone sneaks in some tampered headers, the worst I can see right now is that the rate-limiting won't work.

File: src/ClientBuilder.php
Comment (cleptric) [reply src/ClientBuilder.php:34]:
```php
$httpClient = new class() implements HttpClientInterface {
    public function sendRequest(string $requestData, Options $options): Response
    {
        $client = HttpClient::create([
            'headers' => [
                'Content-Type' => 'application/x-sentry-envelope',
                'User-Agent' => Client::SDK_IDENTIFIER . '/' . Client::SDK_VERSION,
            ],
        ]);

        $response = $client->request('POST', $options->getDsn()->getEnvelopeApiEndpointUrl(), [
            'body' => $requestData,
        ]);

        return new Response($response->getStatusCode(), $response->getHeaders(), '');
    }
};

Sentry\init([
    'dsn' => env('SENTRY_DSN'),
    'http_client' => $httpClient,
]);
```
I like the UX of this approach much more than the boilerplate required today. https://docs.sentry.io/platforms/php/configuration/transports/

We could even go the extra mile and add two sample implementations for Symfony and Guzzle to the SDK.

File: src/HttpClient/HttpClient.php
Comment (Jean85) [reply src/HttpClient/HttpClient.php:130]: That's not true. My was just an example of what security implications we would be opening to if we start writing our own HTTP client.

That specific vulnerability allows the attacker to slip through a custom, malicious body in the response, but I could easily imagine a similar vulnerability in the opposite direction; so you would have the SDK as a vector attack toward our end user or toward Sentry.io or a self-hosted server.

And I'm just limiting myself to the headers. The attack surface may be reduced due to the specific usage of the client (for Sentry only) but it's still there, and it comprises headers, body, format/compression in both requests and responses.

File: src/ClientBuilder.php
Comment (mfb) [reply src/ClientBuilder.php:34]: If it becomes super easy for SDK users to setup the Symfony or Guzzle HTTP client (without boilerplate) then this could be a net improvement :)  I do share concerns that there will be bugs (as well as missing features) when implementing your own HTTP client - we spent a lot of time working on HTTP client issues in the 1.x branch (although to be fair that branch was trying to do more than the 3.x branch, like sending multiple events in parallel at the end of the request).

File: src/HttpClient/HttpClient.php
Comment (stayallive) [inline src/HttpClient/HttpClient.php:83]: ~~I think we should use `\CURLOPT_HEADERFUNCTION` instead. This gives us a callback for each headerline already normalized to lowercase names: https://curl.se/libcurl/c/CURLOPT_HEADERFUNCTION.html. We would also want to set `CURLOPT_SUPPRESS_CONNECT_HEADERS` for the proxy requests.~~

Realized this is a PHP 7.3+ option, so not possible for now but should make header parseing more reliable.

File: src/HttpClient/HttpClient.php
Comment (stayallive) [inline src/HttpClient/HttpClient.php:58]: We might also want to set `CURLOPT_HEADEROPT` to `CURLHEADER_SEPARATE` to be safe. Before cURL `7.42.1` `CURLOPT_HTTPHEADER` was also sent to the proxy which might cause issues, in later versions `CURLHEADER_SEPARATE` became the standard but since we don't know the cURL version we might want to set this ourselfs.

File: src/Client.php
Comment (stayallive) [inline src/Client.php:178]: We might also want to wrap this in a `trace()` so requests to Sentry are logged as a span in case errors are transmitted inside a performance traced request (or maybe that should be done inside the Transport or somewhere else but it would be nice to have a span for it.

File: src/ClientBuilder.php
Comment (stayallive) [inline src/ClientBuilder.php:78]: We should move this down to the `getClient()` which is the actual builder method. And only create a client or use the one from the options if the transport was not set. Also saves needing the `httpClient` instance variable.

File: src/HttpClient/HttpClient.php
Comment (stayallive) [inline src/HttpClient/HttpClient.php:47]: Since we don't actually use the body anywhere we should disable this to save on memory. Headers should still be returned (untested).

File: src/Options.php
Comment (stayallive) [inline src/Options.php:675]: I'm wondering if we should only allow the `http_client` option. Since right now setting a `transport` overshadows the client. If you need a custom transport I don't think it's bad to use the client builder directly but setting the `http_client` seems like a reasonable thing to have especially since the `transport` has quite a few more dependencies needed to work in almost all cases, WDYT?

File: src/Options.php
Comment (stayallive) [inline src/Options.php:981]: Didn't we support proxy authentication before? We might otherwise need to make sure you couldn't pass the `http_proxy` as `user:pass@proxyhost.example` or we have a BC, not a big one, but keeping it simpler might be worth splitting `http_proxy` on the `@` in the client instead of adding another option for it.

File: src/Util/Http.php
Comment (stayallive) [inline src/Util/Http.php:41]: Micro-optim: maybe move the `$headers` down and remove the `array_merge`.

File: src/Util/Http.php
Comment (stayallive) [inline src/Util/Http.php:47]: This might be the most fragile part from what I'm reading online. Not saying I have a better option currently but this requires some more scrutiny than I can give at this moment.

**Review by stayallive — COMMENTED** [review]
Overal this is starting to look great, mostly because it's a heck of a lot simpler.

File: src/HttpClient/HttpClient.php
Comment (Jean85) [reply src/HttpClient/HttpClient.php:83]: Why? Do we still want to support PHP 7.2 and below even in the next major?

Comment (stayallive) [reply src/HttpClient/HttpClient.php:83]: I made a mistake `CURLOPT_HEADERFUNCTION` is available way longer, not starting in PHP 7.3, I've changed my comment to reflect that 🤦

File: src/Client.php
Comment (cleptric) [reply src/Client.php:226]: This could become handy if we ever invest more into async runtimes. I would keep it.

File: src/HttpClient/HttpClient.php
Comment (cleptric) [reply src/HttpClient/HttpClient.php:47]: > **CURLOPT_RETURNTRANSFER** `true` to return the transfer as a string of the return value of `curl_exec()` instead of outputting it directly.

If we remove this, we will just echo the response string. Better to keep it like that 😬

File: src/Options.php
Comment (cleptric) [reply src/Options.php:675]: I'm aware of some people who added custom retry logic and such. I would like to keep this as is, to ease swapping out the transport. The current logic in `ClientBuilder::__construct` should work well, or am I overlooking something?

File: src/Options.php
Comment (cleptric) [reply src/Options.php:981]: Imo, we did not. Referring to https://curl.se/libcurl/c/CURLOPT_PROXY.html

File: src/HttpClient/HttpClient.php
Comment (cleptric) [reply src/HttpClient/HttpClient.php]: I like this a lot, thanks Alex! Have a look at https://github.com/getsentry/sentry-php/pull/1589/commits/c9b3a2da444767b2793c5994a228e92c1675d4c1

File: src/Options.php
Comment (stayallive) [reply src/Options.php:981]: Huh, interesting, okay!

File: src/HttpClient/HttpClient.php
Comment (stayallive) [reply src/HttpClient/HttpClient.php:53]: Yikes! Good call 😉

File: src/Client.php
Comment (stayallive) [reply src/Client.php:178]: Okay that would make sense, although it also doesn't 😅 Interestring to know why since it could account for a large-ish gap in the spans if the app sends an error and the connection is terrible for some reason. Would be great to know (and nice insights maybe for Sentry data-miners).

**Review by stayallive — APPROVED** [review]
👍🚀

**Comment by cleptric** [comment]
Merging now to unblock other features. The remaining ToDos are documented in https://github.com/getsentry/sentry-php/issues/1595

File: src/ClientBuilder.php
Comment (cleptric) [reply src/ClientBuilder.php:78]: Moved to https://github.com/getsentry/sentry-php/issues/1595

File: src/Options.php
Comment (cleptric) [reply src/Options.php:675]: Moved to https://github.com/getsentry/sentry-php/issues/1595

File: src/Client.php
Comment (cleptric) [reply src/Client.php:178]: TBD, will get back later.

**Pull request merged**
